### PR TITLE
Add inline-assembly support

### PR DESCRIPTION
**Pull request by Flakebi**
So far, it is complicated and time-consuming to test the behavior of
certain instructions or their interactions. The easiest method to test
a sequence of instructions to date, is to make a simple graphics test,
disassemble the pipelines, edit the assembly, recompile them and use
pipeline replacement to load the edited code.

This patch adds a simpler and more robust way by letting one write
inline assembly in glsl.

glsl allows strings only in a single place, which is the debug print
instruction. Using the special '%ra' format specifier [0] triggers llpc
to interpret the second argument as an inline assembly string.

The arguments to printf are:
- "%ra" for radeon-assembly
- assembly instructions
- variable constraints
- "no output" or a variable where the output is stored (optional if
  there are no further arguments)
- zero or more arguments

See the llvm reference for more details on how to write the constraint
string: https://llvm.org/docs/LangRef.html#inline-assembler-expressions

The added tests contain an example of how inline assembly can be used.

[0]: The list of existing format specifiers does not include a %r:
https://github.com/KhronosGroup/Vulkan-ValidationLayers/blob/master/docs/debug_printf.md#debug-printf-format-string

### CODE REVIEW

File: test/amber/a16.amber [hunk 1]
```diff
@@ -16,7 +16,7 @@
 # Modifications Copyright (c) 2022 Advanced Micro Devices, Inc. All rights reserved.
 
 # TODO Remove the not once this is fixed in LLVM
-# RUN: not run_amber_test.py --icd %icd %s | FileCheck %s
+# RUN: run_amber_test.py --icd %icd %s | FileCheck %s
```
Comment (jayfoad) [inline test/amber/a16.amber:19]: Is this an unrelated fix?

**Comment by JaxLinAMD** [comment]
retest this please

**Comment by amdvlk-admin** [comment]
### Test summary for commit 4ddc59b2ce2727ce799114b29d1a96e397ee180d
<details><summary>CTS tests (Failed: 0/195808)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 38337/67464 (56.8%)</li>
<li>Failed: 0/67464 (0.0%)</li>
<li>Not Supported: 29127/67464 (43.2%)</li><li>Warnings: 0/67464 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 33521/60880 (55.1%)</li>
<li>Failed: 0/60880 (0.0%)</li>
<li>Not Supported: 27359/60880 (44.9%)</li><li>Warnings: 0/60880 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 39335/67464 (58.3%)</li>
<li>Failed: 0/67464 (0.0%)</li>
<li>Not Supported: 28129/67464 (41.7%)</li><li>Warnings: 0/67464 (0.0%)</li>
</ul></details>


</ul></details>

File: llpc/translator/lib/SPIRV/SPIRVReader.cpp [hunk 2]
```diff
@@ -3778,9 +3779,90 @@ template <> Value *SPIRVToLLVM::transValueWithOpcode<OpGroupFMaxNonUniformAMD>(S
 template <> Value *SPIRVToLLVM::transValueWithOpcode<OpExtInst>(SPIRVValue *const spvValue) {
   SPIRVExtInst *const spvExtInst = static_cast<SPIRVExtInst *>(spvValue);
 
-  // Just ignore this set of extended instructions
-  if (m_bm->getBuiltinSet(spvExtInst->getExtSetId()) == SPIRVEIS_NonSemanticInfo)
-    return nullptr;
+  if (m_bm->getBuiltinSet(spvExtInst->getExtSetId()) == SPIRVEIS_NonSemanticInfo) {
+    switch (spvExtInst->getExtOp()) {
+    case 1: { // DebugPrintf
```
Comment (amdrexu) [inline llpc/translator/lib/SPIRV/SPIRVReader.cpp:3784]: I think we ought to import the file to LLPC like what we have done for GLSL.std.450.h. Of course, this extended instruction set has just one instruction currently but maybe it will be enlarged in the future.
https://github.com/KhronosGroup/SPIRV-Headers/blob/master/include/spirv/unified1/NonSemanticDebugPrintf.h

File: llpc/translator/lib/SPIRV/SPIRVReader.cpp [hunk 3]
```diff
@@ -3778,9 +3779,90 @@ template <> Value *SPIRVToLLVM::transValueWithOpcode<OpGroupFMaxNonUniformAMD>(S
...
+      auto &spvArgs = spvExtInst->getArguments();
+      assert(spvArgs.size() >= 1 && "printf must have at least one argument");
+      auto &formatString = m_bm->get<SPIRVString>(spvArgs[0])->getStr();
+      // Ignore printfs that don't have a matching format specifiers
+      if (formatString.find("%r") == std::string::npos)
+        return nullptr;
+
+      if (formatString.find("%ra") != std::string::npos) {
```
Comment (amdrexu) [inline llpc/translator/lib/SPIRV/SPIRVReader.cpp:3792]: I think we are free to use this extended instruction to implement our vendor-specific debugging. So I prefer to add relevant codes to a sperate handler of DebugPrintf and documents its usage with descriptive comments like `--help` print.

**Comment by Flakebi** [comment]
V2: Include SPIR-V header, move code into a function and use llvm diagnostics for errors so that it also works in release mode.

Still missing some documentation (is there a better place for documenting this than the comment in SPIRVReader.cpp?).

**Review by amdrexu — DISMISSED** [review]
If you feed it is necessary, you can add a brief doc. But currently, the functionality seems straightforward by reading the inline comments.

**Comment by amdvlk-admin** [comment]
### Test summary for commit e8b0a6742b3d0209243b47da51e36ac5ec57b171
<details><summary>CTS tests (Failed: 0/195808)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 38337/67464 (56.8%)</li>
<li>Failed: 0/67464 (0.0%)</li>
<li>Not Supported: 29127/67464 (43.2%)</li><li>Warnings: 0/67464 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 33521/60880 (55.1%)</li>
<li>Failed: 0/60880 (0.0%)</li>
<li>Not Supported: 27359/60880 (44.9%)</li><li>Warnings: 0/60880 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 39335/67464 (58.3%)</li>
<li>Failed: 0/67464 (0.0%)</li>
<li>Not Supported: 28129/67464 (41.7%)</li><li>Warnings: 0/67464 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdrexu** [comment]
@Flakebi There are some internal discussions about this PR. Please check them with Nicolai. The porting is blocked.
